src/ts/imageloader.ts

```diff
@@ -0,0 +1,80 @@
+import {DOM} from './dom';
+
+export interface ImageLoadedCallback {
+  (url: string, width: number, height: number): void;
+}
+
+interface ImageLoaderState {
+  url: string,
+  image: DOM;
+  loadedCallback: ImageLoadedCallback;
+  loaded: boolean;
+  width: number;
+  height: number;
+}
+
+/**
+ * Tracks the loading state of images.
+ */
+export class ImageLoader {
+
+  private state: { [url: string]: ImageLoaderState; } = {};
+
+  /**
+   * Loads an image and call the callback once the image is loaded. If the image is already loaded, the callback
+   * is called immediately, else it is called once loading has finished. Calling this method multiple times for the
+   * same image while it is loading calls only let callback passed into the last call.
+   * @param url The url to the image to load
+   * @param loadedCallback The callback that is called when the image is loaded
+   */
+  load(url: string, loadedCallback: ImageLoadedCallback): void {
+    if (!this.state[url]) {
+      // When the image was never attempted to be loaded before, we create a state and store it in the state map
+      // for later use when the same image is requested to be loaded again.
+      let state: ImageLoaderState = {
+        url: url,
+        image: new DOM('img', {}),
+        loadedCallback: loadedCallback,
+        loaded: false,
+        width: 0,
+        height: 0,
+      };
+      this.state[url] = state;
+
+      // We need to add the image element to the DOM to get the size of the loaded image
+      new DOM('body').append(state.image);
```

Comment:
I think we should set the `display` of the img-tag to `none` before appending it to the body to make sure that it is never visible to the user, even when it is directly removed in `onload`.